storage/storage.go

```diff
@@ -90,55 +90,132 @@ func (storage *Storage) getAllChunks(context context.Context, buildId string) ([
+	if err != nil {
+		return nil, err
+	}
+
+	sortByStartTime(buildChunks)
+	sortByStartTime(testChunks)
+
+	timeRange := TimeRange{
+		StartAt: minTime(buildChunks[0].Start, testChunks[0].Start),
+		EndAt:   maxTime(getLatestTime(buildChunks), getLatestTime(testChunks)),
+	}
+
+	buildChunkIterator := NewBatchedLogIterator(storage.bucket, buildChunks, 4, timeRange)
+	testChunkIterator := NewBatchedLogIterator(storage.bucket, testChunks, 4, timeRange)
+
+	// Merge test and build logs
+	return NewMergingIterator(testChunkIterator, buildChunkIterator), nil
+}
+
+func testChunksWithId(chunks []LogChunkInfo, testID string) []LogChunkInfo {
+	testChunks := []LogChunkInfo{}
 	for i := 0; i < len(chunks); i++ {
-		// Include any build logs that are in the time range of our test
-		chunkTimeRange := TimeRange{
-			StartAt: chunks[i].Start,
-			EndAt:   chunks[i].End,
+		// Find our test id
+		if chunks[i].TestID == testID {
+			testChunks = append(testChunks, chunks[i])
 		}
-		// check if the global build chunk's time range intersects the test's time range, and if so
-		// add it to our list of build chunks, but constrained to the test's time range to only
-		// include entries during that time.
-		if chunks[i].TestID == "" && testTimeRange.Intersects(chunkTimeRange) {
-			buildChunks = append(buildChunks, chunks[i])
+	}
+	return testChunks
+}
+
+// Gets the first chunk in the list with a start time after "target", otherwise falls back to the
+// fallback time if there are no such chunks.
+func getFirstTestChunkAfter(allTestChunks []LogChunkInfo, target time.Time, fallback time.Time) time.Time {
+	for _, chunk := range allTestChunks {
+		if chunk.Start.After(target) {
+			return chunk.Start
 		}
 	}
+	return fallback
+}
 
-	sort.Slice(buildChunks, func(i, j int) bool {
-		return buildChunks[i].Start.Before(buildChunks[j].Start)
-	})
+func (storage *Storage) GetTestLogLines(context context.Context, buildId string, testId string) (LogIterator, error) {
+	buildChunks, allTestChunks, err := storage.getBuildAndTestChunks(context, buildId)
+	if err != nil {
+		return nil, err
+	}
+
+	sortByStartTime(allTestChunks)
+
+	testChunks := testChunksWithId(allTestChunks, testId)
+
+	sortByStartTime(testChunks)
+
+	// We want to get all logs up to the next test chunk after the chunks in our queried test.
+	// If there are no test chunks after our queried test, then this should set the end time
+	// to the latest end time of all the chunks we have.
+	lastTestChunkEnd := getLatestTime(testChunks)
+	maxPossibleTime := maxTime(getLatestTime(buildChunks), lastTestChunkEnd)
+	logEndTime := getFirstTestChunkAfter(allTestChunks, lastTestChunkEnd, maxPossibleTime)
+
+	testTimeRange := TimeRange{
+		StartAt: testChunks[0].Start,
+		EndAt:   logEndTime,
+	}
+
+	testChunkIterator := NewBatchedLogIterator(storage.bucket, testChunks, 4, testTimeRange)
+
+	sortByStartTime(buildChunks)
+	// This batchedlogiterator will filter out buildChunks that don't intestect with testTimeRange
```

Comment:
"intestect" -> "intersect"

Reply:
Although they will be filtered out it seems like it could be more efficient if we only pass chunks that intersect with the time range, at least for later tests in the build, since the iterator gets each chunk before it filters it out.

Reply:
They get pre-filtered [here](https://github.com/evergreen-ci/logkeeper/blob/main/storage/log_iterator.go#L199) by the iterator, before fetching anything.

Reply:
I will fix the comment though.

Reply:
oh! I missed that